### PR TITLE
Keep the order of command line arguments for build.

### PR DESCRIPTION
**Pull request by ocallesp**
What are the changes:
Get the command line arguments from IProjectBuildSnapshotService instead of IProjectSubscriptionService.

Subscribe to the ProjectBuildSnapshotService directly using the build targets, instead of rules.

What is the bug:
Getting the data from the rules changes the order of the commands which causes the Build to behave differently.

How this change fixes the bug:
Get the original order from ProjectBuildSnapshotService and pass it directly to roslyn.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/7174)

### CODE REVIEW

**Comment by ocallesp** [comment]
@drewnoakes this reverts the previous fix, and also make the changes as you suggested.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs [hunk 1]
```diff
@@ -146,19 +145,12 @@ private void ProcessProjectBuildFailure(IProjectRuleSnapshot snapshot)
...
-            // The order of options matters and we should use warnaserror- and then warnaserror+
-            // Depending which comes first, the compiler produces different diagnostics.
-            // We just need to move warnaserror+ to the end.
-            var warnaserrorPlus = snapshot.Items.FirstOrDefault(a => a.Key.Contains("/warnaserror+"));
-            IEnumerable<string> commandLineArguments = warnaserrorPlus.Key is null ? snapshot.Items.Keys :
-                    snapshot.Items.Remove(warnaserrorPlus.Key).Keys.Append(warnaserrorPlus.Key);
-
             // We just pass all options to Roslyn
-            _context.SetOptions(commandLineArguments.ToImmutableArray());
+            _context.SetOptions(arguments.ToImmutableArray());
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:153]: Please do the revert in a separate PR, to make this easier to review.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs [hunk 2]
```diff
@@ -103,6 +108,16 @@ protected override async Task InitializeCoreAsync(CancellationToken cancellation
                             linkOptions: DataflowOption.PropagateCompletion,
                             cancellationToken: cancellationToken),
 
+                    ProjectDataSources.SyncLinkTo(
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs:111]: This needs to be sync-linked with the ProjectBuildRuleSource, otherwise, you are going to be pushing changes to Roslyn that are out of sync in version. ie The command-line arguments will be out of sync with the other changes we're pushing to Roslyn, such as source files, references, etc.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/IApplyChangesToWorkspaceContext.cs [hunk 3]
```diff
@@ -83,5 +83,14 @@ internal interface IApplyChangesToWorkspaceContext
...
         Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> update, ContextState state, CancellationToken cancellationToken);
+
+        /// <summary>
+        ///     Sets project build command line options to the underlying <see cref="IWorkspaceProjectContext"/>.
+        /// </summary>
+        /// <param name="arguments">List of command line arguments in the Build</param>
+        /// <remarks>
+        ///     Note: The compiler is sensitive to ordering of command-line arguments.
+        /// </remarks>
+        void SetCommandLineArgumentsToBuild(IEnumerable<string> arguments);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/IApplyChangesToWorkspaceContext.cs:94]: Rather than adding a new member here, it'd be better to pass the project build snapshot as part of `ApplyProjectBuildAsync`. If you look at `ApplyProjectChangesUnderLockAsync` you will see it is called within a lock, and does a bunch of other scheduling (eg. start/end batch). Making another method here and calling it out of sync with `ApplyProjectBuildAsync` will lead to data races and potentially other kinds of issues.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs [hunk 4]
```diff
@@ -103,6 +108,16 @@ protected override async Task InitializeCoreAsync(CancellationToken cancellation
...
+                        _activeConfiguredProjectProvider.ActiveConfiguredProjectBlock.SyncLinkOptions(),
+                        _projectBuildSnapshotService.SourceBlock.SyncLinkOptions(),
+                        target: DataflowBlockFactory.CreateActionBlock<IProjectVersionedValue<ValueTuple<ConfiguredProject, IProjectBuildSnapshot>>>(e =>
+                                ApplyBuildSnapshotChanged(e),
+                            _project.UnconfiguredProject,
+                            ProjectFaultSeverity.LimitedFunctionality),
+                        linkOptions: DataflowOption.PropagateCompletion,
+                        cancellationToken: cancellationToken),
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs:119]: Instead of creating a new link here, add the project build snapshot into the above build link.

The new data source must be sync-linked with other sources to avoid races.

`OnProjectChangedAsync` is called for both evaluation and build results. You will likely need to add another member to the `ValueTuple` for the new data, and have it as `null` for evaluation data. Then you can pass the `IProjectBuildSnapshot` through to `ApplyProjectBuildAsync`.

**Review by drewnoakes — CHANGES_REQUESTED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs [hunk 5]
```diff
@@ -162,6 +177,20 @@ public async Task<T> OpenContextForWriteAsync<T>(Func<IWorkspaceProjectContextAc
                 }
             }
 
+            private void ApplyBuildSnapshotChanged(IProjectVersionedValue<(ConfiguredProject project, IProjectBuildSnapshot buildSnapshot)> update)
+            {
+                List<string> commandLineArguments = new List<string>();
+
+                update.Value.buildSnapshot.TargetOutputs.TryGetValue("CompileDesignTime", out var keyValuePairsCommandLineOptions);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs:184]: Please make this way more efficient; as spoken about earlier this is very sensitive to allocations and LOH allocations, the List above will be resized many times over. I think we want a more efficient way to go from the IEnumerable<string> -> ImmutableArray.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/IApplyChangesToWorkspaceContext.cs (hunk 3 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/IApplyChangesToWorkspaceContext.cs]: Done, please latest commit "Sync-link buildSnapshot with the ProjectBuildRuleSource."

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs (hunk 1 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:175]: main already has the revert.

This PR is a rebase on main having the revert.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs (hunk 2 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs]: Done. This is implemented in the commit "Sync-link buildSnapshot with the ProjectBuildRuleSource"

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs [hunk 6]
```diff
@@ -57,7 +58,10 @@ public void Initialize(IWorkspaceProjectContext context)
             EnsureInitialized();
         }
 
-        public async Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> update, ContextState state, CancellationToken cancellationToken)
+        public async Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> update, 
+            IProjectVersionedValue<IProjectBuildSnapshot> buildSnapshot,
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:62]: This value does not need to be versioned, which will simplify calling code slightly.

```suggestion
            IProjectBuildSnapshot buildSnapshot,
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs [hunk 7]
```diff
@@ -69,11 +73,29 @@ public async Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscrip
...
+
+            buildSnapshot.TargetOutputs.TryGetValue("CompileDesignTime", out var keyValuePairsCommandLineOptions);
+
+            foreach (var kp in keyValuePairsCommandLineOptions)
+            {
+                commandLineArguments.Add(kp.Key);
+            }
+
+            // We just need to pass all options to Roslyn
+            _context.SetOptions(commandLineArguments.ToImmutableArray());
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:96]: To reduce allocations (written in a browser, so untested):

```suggestion
            buildSnapshot.TargetOutputs.TryGetValue(
                "CompileDesignTime",
                out IImmutableList<KeyValuePair<string, IImmutableDictionary<string, string>>> targetOutputs);

            var options = ImmutableArray.CreateBuilder<string>(targetOutputs.Count);

            foreach ((string option, _) targetOutputs)
            {
                options .Add(option);
            }

            // We just need to pass all options to Roslyn
            _context.SetOptions(options .MoveToImmutable());
```

By using a builder with a predefined capacity, and adding exactly that number of items, we can use `MoveToImmutable` which moves the backing array from the builder to the `ImmutableArray<string>`. This means there is only a single array allocated during this operation.

You need to handle the case where `TryGetValue` returns false, as then `targetOutputs` will be `null`, resulting in an exception.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs [hunk 8]
```diff
@@ -177,32 +168,19 @@ public async Task<T> OpenContextForWriteAsync<T>(Func<IWorkspaceProjectContextAc
...
 
-            private async Task ApplyProjectChangesUnderLockAsync(IProjectVersionedValue<(ConfiguredProject project, IProjectSubscriptionUpdate subscription)> update, bool evaluation, CancellationToken cancellationToken)
+            private async Task ApplyProjectChangesUnderLockAsync(IProjectVersionedValue<(ConfiguredProject project, IProjectSubscriptionUpdate subscription, IProjectBuildSnapshot buildSnapshot)> update, bool evaluation, CancellationToken cancellationToken)
             {
                 // NOTE we cannot call CheckForInitialized here, as this method may be invoked during initialization
                 Assumes.NotNull(_contextAccessor);
 
                 IWorkspaceProjectContext context = _contextAccessor.Context;
                 IProjectVersionedValue<IProjectSubscriptionUpdate> subscription = update.Derive(u => u.subscription);
+                IProjectVersionedValue<IProjectBuildSnapshot> buildSnapshot = update.Derive(u => u.buildSnapshot);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs:183]: As mentioned earlier, we don't need a versioned value for the build snapshot, so we can avoid the `Derive` here and reduce allocations.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs (hunk 5 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs]: done

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs (hunk 7 above)
Comment (MiYanni) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs]: If you know the size beforehand, what is the advantage of using an `ImmutableArray<T>.Builder` over just using a normal array, populating it, and then calling `ImmutableArray<T>.Create()`? It seems more straightforward that way.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs [hunk 9]
```diff
@@ -69,12 +73,31 @@ public async Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscrip
...
 
+        private void ProcessOptions(IProjectBuildSnapshot buildSnapshot)
+        {
+            Assumes.NotNull(_context);
+
+            buildSnapshot.TargetOutputs.TryGetValue(
+                "CompileDesignTime",
+                out IImmutableList<KeyValuePair<string, IImmutableDictionary<string, string>>> targetOutputs);
+
+            var options = ImmutableArray.CreateBuilder<string>(targetOutputs.Count);
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:90]: What makes TargetsOuputs.Count a good capacity here?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs [hunk 10]
```diff
@@ -146,19 +169,12 @@ private void ProcessProjectBuildFailure(IProjectRuleSnapshot snapshot)
             }
         }
 
-        private void ProcessOptions(IProjectRuleSnapshot snapshot)
+        public void SetCommandLineArgumentsToBuild(IEnumerable<string> arguments)
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:172]: When is this method called?

**Comment by davkean** [comment]
This looks much better. Please make sure you test this heavily and ensure that the bug you are fixing is fixed, plus we didn't regress passing options to Roslyn (such as Version, conditional compilation symbols, etc).

**Review by davkean — APPROVED** [review]
Assuming SetCommandLineArgumentsToBuild is being removed, this looks good.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs (hunk 8 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/WorkspaceProjectContextHost.WorkspaceProjectContextHostInstance.cs]: this was fixed, this is the code cleanup

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs (hunk 9 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:90]: Yes, in order to `Move` we need to get the count right 100% of the time, otherwise it'll throw. There isn't a more efficient way than this.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs [hunk 11]
```diff
@@ -69,12 +73,31 @@ public async Task ApplyProjectBuildAsync(IProjectVersionedValue<IProjectSubscrip
...
+
+            var options = ImmutableArray.CreateBuilder<string>(targetOutputs.Count);
+
+            foreach ((string option, _) in targetOutputs)
+            {
+                options.Add(option);
+            }
+
+            // We just need to pass all options to Roslyn
+            _context.SetOptions(options .MoveToImmutable());
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs:98]:
```suggestion
            _context.SetOptions(options.MoveToImmutable());
```

**Review by drewnoakes — APPROVED** [review]
Agree that `SetCommandLineArgumentsToBuild` looks like it can be removed. Other than that, this looks like a good change (I haven't tested it locally).

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs (hunk 10 above)
Comment (ocallesp) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs]: this method was removed

**Comment by ocallesp** [comment]
cc. @jcouv

**Comment by jcouv** [comment]
Awesome. Thanks much!
FYI @RikkiGibson

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs (hunk 7 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/ApplyChangesToWorkspaceContext.cs]: @MiYanni I never replied.

`ImmutableArray<T>.Create(T[])` will copy the contents of the passed array to a new array, resulting in extra work and allocation. It does this copy so that you cannot modify the contents of the array. `ImmutableArray<>` is a wrapper around an array, and it is designed to make it very hard to modify the contents of the array, as the type's name would suggest.

The builder can encapsulate the array, so no one outside can modify it without its knowledge (via its API). The `MoveToImmutable` method "moves" the builder's array directly into the `ImmutableArray<>`.
